sui-execution/latest/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -1502,6 +1549,64 @@ mod checked {
         }
     }
 
+    /// Pushes event to the trace builder if tracing is enabled.
+    /// Event is created via `create_event` function taking TypeTagWithRefs
+    /// as an argument that is created from `type_` inside this function.
+    /// This somewhat complicated code structure was introduced to make sure
+    /// that the invariant violation that can result from tag creation failure
+    /// will only trigger when tracing is enabled.
+    fn push_trace_event_with_type_tag(
+        context: &mut ExecutionContext<'_, '_, '_>,
+        type_: &Type,
+        trace_builder_opt: &mut Option<MoveTraceBuilder>,
+        create_event: impl FnOnce(TypeTagWithRefs) -> TraceEvent,
+    ) -> Result<(), ExecutionError> {
+        if let Some(trace_builder) = trace_builder_opt {
+            let (deref_type, ref_type) = match type_ {
+                Type::Reference(t) => (t.as_ref(), Some(RefType::Imm)),
+                Type::MutableReference(t) => (t.as_ref(), Some(RefType::Mut)),
+                t => (t, None),
+            };
+            // this invariant violation will only trigger when tracing
+            let type_tag = context
+                .vm
+                .get_runtime()
+                .get_type_tag(deref_type)
+                .map_err(|e| {
+                    ExecutionError::new_with_source(ExecutionErrorKind::InvariantViolation, e)
+                })?;
+            let type_tag_with_ref = TypeTagWithRefs {
+                type_: type_tag,
+                ref_type,
+            };
+            trace_builder.push_event(create_event(type_tag_with_ref));
+        }
+        Ok(())
+    }
+
+    /// Converts a type to type tag format used in tracing.
+    /// SHOULD ONLY BE USED WHEN TRACING IS ENABLED as it may
+    /// cause invariant violation.
+    fn trace_type_to_type_tag_with_refs(
```

Comment:
This is currently unused in favor of `push_trace_event_with_type_tag` which I feel more comfortable using as it helps asserting that the code is only executed when tracing (but perhaps it's a bit of an overkill)